internal/watcher/minio.go

```diff
@@ -81,42 +81,45 @@ func NewMinioWatcher(ctx context.Context, logger logr.Logger, config *MinioConfi
 	}, nil
 }
 
-func (w *minioWatcher) Watch(ctx context.Context) (*BlobEvent, error) {
-	event, err := w.pop(ctx)
+func (w *minioWatcher) Watch(ctx context.Context) (*BlobEvent, Cleanup, error) {
+	event, val, err := w.pop(ctx)
+	clf := func() error {
+		if _, err := w.client.LRem(ctx, w.failedList, 1, val).Result(); err != nil {
+			w.logger.Error(err, "Error removing message from failed list.", "list", w.failedList)
+			return err
+		}
+		w.logger.V(2).Info("Successfully removed messages.")
+
+		return nil
+	}
 	if err != nil {
 		if errors.Is(err, redis.Nil) {
 			err = ErrWatchTimeout
 		}
-		return nil, err
+		return nil, clf, err
 	}
 	if event.Bucket != w.bucket {
-		return nil, ErrBucketMismatch
+		return nil, clf, ErrBucketMismatch
 	}
-	return event, nil
+	return event, clf, nil
 }
 
 func (w *minioWatcher) Path() string {
 	return ""
 }
 
-func (w *minioWatcher) pop(ctx context.Context) (*BlobEvent, error) {
+func (w *minioWatcher) pop(ctx context.Context) (*BlobEvent, string, error) {
 	val, err := w.client.BLMove(ctx, w.listName, w.failedList, "RIGHT", "LEFT", redisPopTimeout).Result()
 	if err != nil {
-		return nil, fmt.Errorf("error retrieving from Redis list: %w", err)
+		return nil, "", fmt.Errorf("error retrieving from Redis list: %w", err)
 	}
 
 	event, err := w.event(val)
 	if err != nil {
-		return nil, fmt.Errorf("error processing item received: %w", err)
-	}
-
-	if _, err := w.client.LRem(ctx, w.failedList, 1, val).Result(); err != nil {
-		w.logger.Error(err, "Error removing message from failed list.", "list", w.failedList)
-	} else {
-		w.logger.V(2).Info("Successfully removed messages.")
+		return nil, "", fmt.Errorf("error processing item received: %w", err)
 	}
 
-	return event, nil
+	return event, val, nil
 }
```

Comment:
I haven't tested but if I remember correctly you could also have a method match the Cleanup function signature, so the cleanup can be a `w.rem` method just like `w.pop` is a method too:

```go
func (w *minioWatcher) Watch(ctx context.Context) (*BlobEvent, Cleanup, error) {
	event, val, err := w.pop(ctx)
	if err != nil {
		if errors.Is(err, redis.Nil) {
			err = ErrWatchTimeout
		}
		return nil, noopCleanup, err
	}

	if event.Bucket != w.bucket {
		return nil, noopCleanup, ErrBucketMismatch
	}

	return event, w.rem, nil
}


func (w *minioWatcher) rem(ctx context.Context, val string) error {
	logger := w.logger.WithValues("list", w.failedList)

	_, err := w.client.LRem(ctx, w.failedList, 1, val).Result()
	if err != nil {
		logger.Error(err, "Error removing message from failed list.")
		return err
	}
	logger.V(2).Info("Successfully removed message from failed list.")

	return nil
}
```

Reply:
I return a closure instead now